code/modules/wiremod/components/action/pathfind.dm

```diff
@@ -0,0 +1,123 @@
+	id_card = add_input_port("ID Card", PORT_TYPE_ATOM, FALSE)
+
+	output = add_output_port("Next step", PORT_TYPE_ATOM)
+	finished = add_output_port("Arrived to destination", PORT_TYPE_SIGNAL)
+	no_path = add_output_port("Failed: Can't pathfind there", PORT_TYPE_SIGNAL)
+	cooldown = add_output_port("Failed: Cooldown active", PORT_TYPE_SIGNAL)
+	bad_id = add_output_port("Failed: ID not valid", PORT_TYPE_SIGNAL)
+
+/obj/item/circuit_component/pathfind/Destroy()
+	input_X = null
+	input_Y = null
+	id_card = null
+
+	output = null
+	finished = null
+	no_path = null
+	bad_id = null
+
+	path = null
+	old_dest = null
+	next_turf = null
+	return ..()
+
+/obj/item/circuit_component/pathfind/input_received(datum/port/input/port)
+	. = ..()
+	if(.)
+		return
+
+	var/target_X = input_X.input_value
+	if(isnull(target_X))
+		return
+
+	var/target_Y = input_Y.input_value
+	if(isnull(target_Y))
+		return
+
+	var/atom/path_id = id_card.input_value
+	if(path_id && !istype(path_id, /obj/item/card/id))
+		path_id = null
+		bad_id.set_output(COMPONENT_SIGNAL)
+		return
+
+	/// Get both the current turf and the destination's turf
+	var/turf/current_turf = get_turf(src)
+	var/turf/destination = locate(target_X, target_Y, current_turf?.z)
+
+	/// If we're going to the same place and the cooldown hasn't subsided, we're probably on the same path as before
+	if (destination == old_dest && TIMER_COOLDOWN_CHECK(parent, COOLDOWN_CIRCUIT_PATHFIND_SAME))
+
+		/// Check if the current turf is the same as the current turf we're supposed to be in. If so, then we set the next step as the next turf on the list
+		if(current_turf == next_turf)
+			path.Remove(path[1])
+			next_turf = get_turf(path[1])
+			output.set_output(next_turf)
+
+			/// Restart the cooldown since we don't need a new path ( TIMER_COOLDOWN_START might restart the timer by itself and i dont need to call TIMER_COOLDOWN_END, but better safe than sorry )
+			TIMER_COOLDOWN_END(parent, COOLDOWN_CIRCUIT_PATHFIND_SAME)
+			TIMER_COOLDOWN_START(parent, COOLDOWN_CIRCUIT_PATHFIND_SAME, same_path_cooldown)
+
+
+	else /// Either we're not going to the same place or the cooldown is over. Either way, we need a new path
+
+		if(destination != old_dest && TIMER_COOLDOWN_CHECK(parent, COOLDOWN_CIRCUIT_PATHFIND_DIF))
+			cooldown.set_output(COMPONENT_SIGNAL)
+			return
+
+		TIMER_COOLDOWN_END(parent, COOLDOWN_CIRCUIT_PATHFIND_SAME)
+
+		old_dest = destination
+		path = get_path_to(src, destination, 60, id=path_id)
+		if(length(path) == 0 || !path)/// Check if we can even path there
+			no_path.set_output(COMPONENT_SIGNAL)
+			return
+		else
+			TIMER_COOLDOWN_START(parent, COOLDOWN_CIRCUIT_PATHFIND_DIF, different_path_cooldown)
+			path.Remove(path[1]) /// The first step is literally where we are right now, so we dont need it
+			next_turf = get_turf(path[1])
+			output.set_output(next_turf)
+		TIMER_COOLDOWN_START(parent, COOLDOWN_CIRCUIT_PATHFIND_SAME, same_path_cooldown)
+
+	current_turf = get_turf(src)
```

Comment:
current_turf shouldn't really be any different. You can remove this and move the code below up to the start of the code.